source/initial-dns-seedlist-discovery/initial-dns-seedlist-discovery.rst

```diff
@@ -61,35 +64,43 @@ format is::
 specification following the ``Host Information``. This includes the ``Auth
 database`` and ``Connection Options``.
 
-Seedlist Discovery
-------------------
 
-In this preprocessing step, the driver will query the DNS server for SRV
-records on ``{hostname}.{domainname}``, prefixed with the SRV service name
-and protocol. The SRV service name is provided in the ``srvServiceName`` URI option and
-defaults to ``mongodb``. The protocol is always ``tcp``. After prefixing, the URI
-should look like: ``_{srvServiceName}._tcp.{hostname}.{domainname}``. This DNS query
-is expected to respond with one or more SRV records. The driver MUST add all the host
-names and port numbers that were returned as part of the DNS SRV query result to the
-seedlist.
+MongoClient Configuration
+-------------------------
 
-The priority and weight fields in returned SRV records MUST be ignored.
+srvMaxHosts
+~~~~~~~~~~~
 
-If ``mongodb+srv`` is used, a driver MUST implicitly also enable TLS. Clients
-can turn this off by passing ``ssl=false`` in either the Connection String,
-or options passed in as parameters in code to the MongoClient constructor (or
-equivalent API for each driver), but not through a TXT record (discussed in
-the next section).
+This option may be used to limit the number of SRV records used to populate the
+seedlist during initial discovery. It is also used to limit the number of
+additional hosts that may be added during
+`SRV polling <../polling-srv-records-for-mongos-discovery/polling-srv-records-for-mongos-discovery.rst>`_.
+This option requires a non-negative integer and defaults to zero (i.e. no
+limit). This option MUST only be configurable at the level of a ``MongoClient``.
+Drivers MUST report an error if this option is specified and the scheme is not
+``mongodb+srv``.
 
-A driver MUST verify that in addition to the ``{hostname}``, the
-``{domainname}`` consists of at least two parts: the domain name, and a TLD.
-Drivers MUST raise an error and MUST NOT contact the DNS server to obtain SRV
-(or TXT records) if the full URI does not consists of at least three parts.
+This option is used to limit the number of mongos connections that may be
+created for sharded topologies.
```

Comment:
Should we perform any validation of URI options after resolution of the TXT record? Options that seem to be incompatible with `srvMaxHosts` include:

- `loadBalanced=true`
- `replicaSet` (see comment 1 below)
- `directConnection`

I would expect the driver to throw an exception if any of those incompatible options were present along with `srvMaxHosts`.

_comment 1_ It was mentioned that checking for `srvMaxHosts` and `replicaSet` was more trouble than it is worth. Maybe I'm missing some subtlety, but why couldn't we verify that `srvMaxHosts` and `replicaSet` aren't present in the URI and/or TXT record?

Reply:
For context, the only options currently supported in a TXT record are `authSource`, `replicaSet`, and `loadBalanced`.

I think we could still end up connecting to a replica set topology even in the absence of a `replicaSet` URI option. Omitting that option just omits replica set name validation and initializes the topology to an Unknown state. From there, it can still transition based on the servers that are discovered.

`directConnection=true` is already prohibited for `mongodb+srv` schemes, so I don't think we need to do anything there. `directConnection=false` is not a logical conflict for SRV, and I believe the URI options spec was already amended to allow that.

That leaves `loadBalanced=true` (again, `false` is not an issue). In that case, the topology type won't be sharded so `srvMaxHosts` will have no effect (similar to using it with a replica set).

If you or anyone else is strongly in favor of this validation I can add it. I reckon we would just prohibit `srvMaxHosts` it for `loadBalanced=true` and the presence of a `replicaSet` URI option, and do nothing if the topology later becomes a replica set after SDAM kicks in.

Reply:
Raising an error for conflicting URI args seems like a small ask and easy to validate. I'll make the change and add a few tests. Will also explicitly note that we'll ignore the edge case where `replicaSet` is omitted but the user ends up connecting to a replica set topology anyway.